core/kernel/classes/class.Resource.php

```diff
@@ -187,40 +176,32 @@ public function getTypes()
      * method preventing to call the get getPropertyValues() method for a such common
      * operation.
      *
-     * @author Patrick Plichart <patrick@taotesting.com>
      * @return string A Uniform Resource Identifier (URI).
+     * @throws core_kernel_persistence_Exception
+     * @author Patrick Plichart <patrick@taotesting.com>
      */
-    public function getLabel()
-    {
-        if (is_null($this->label)) {
-            $label =  $this->getOnePropertyValue($this->getProperty(OntologyRdfs::RDFS_LABEL));
-            $this->label = is_null($label)
-                ? ''
-                : ($label instanceof core_kernel_classes_Resource
-                    ? $label->getUri()
-                    : (string)$label
-                   )
-            ;
+    public function getLabel(): ?string
+    {
+        $label =  $this->getOnePropertyValue($this->getProperty(OntologyRdfs::RDFS_LABEL));
+
+        if ($label === null) {
+            return '';
         }
-        
-        return $this->label;
+
+        return $label instanceof self ? $label->getUri() : __((string)$label);
```

Comment:
I am afraid that it can break some translations like `__('%s Class saved', $clazz->getLabel())` cause label can be translated and we don't have a translation for half translated text

also we need to care of cases like `__\(\$.*->getLabel` for example `__($property->getLabel())`

Reply:
As an option, I can add a parameter to `getLabel` with the default option, and if need to turn off translation, the code user can configure it during the call. Something like this:
```
public function getLabel(bool $translate = true) {}
// ...
$v = __($property->getLabel(false));
```

Reply:
As it already breaks BC, adding a new property with default option seems pretty safe.

Reply:
default `true` will do the same. it will break a lot of translations.
but default `false` is not better than __($property->getLabel()). you still need to go through all extension and put this `true` 
anyway, I believe this is not a responsibility of the Resource class to translate labels complexity and coupling getting higher.

Reply:
Translating the label should be restricted on TAO internal resources, not on all resources. 
Example: define a new class with Label = Label or any other keyword we translate. It will get translated as well. We need to  prohibit this behavior.

Reply:
We already have a system in place to translate labels, please check how the column "l_language" is used, and this can work on all resources, even user defined resources.

Please do not use the code translation tool on top of the ontology translation, this could have very weird effects

Reply:
Also @SergiiTao this is not how __('%s Class saved', $clazz->getLabel()) works. The translation is applied before the string replacement.

Users are asked to translate '%s Class saved' so it will work with any class label. Please keep using this format and never use construct such as:
`__($clazz->getLabel() . ' Class saved')`
or 
`$clazz->getLabel() . __(' Class saved')`